packages/react-aria-components/src/DropZone.tsx

```diff
@@ -0,0 +1,178 @@
+
+export const DropZoneContext = createContext<ContextValue<DropZoneProps, HTMLDivElement>>(null);
+
+function DropZone(props: DropZoneProps, ref: ForwardedRef<HTMLDivElement>) {
+  [props, ref] = useContextProps(props, ref, DropZoneContext);
+  let {onPress, onPressStart, onPressEnd, onPressChange, onPressUp} = props;
+  let {dropProps, isDropTarget} = useDrop({...props, ref});
+  let {hoverProps, isHovered} = useHover({});
+  let {focusProps, isFocused, isFocusVisible} = useFocusRing({within: true});
+
+  let buttonRef = useRef<HTMLButtonElement>(null);
+  let inputRef = useRef<HTMLInputElement>(null);
+  let hasButton = useHasChild('button[slot=file]', ref);
+  let hasLink = useHasChild('span[slot=file]', ref);
+
+  let hasInput = useHasChild('input[type=file]', ref);
+  let textId = useId();
+  let isVirtualDrag = isVirtualDragging();
+  let labelText = useMemo(() => {
+    if (!isVirtualDrag && hasInput) {
+      return 'Open file system';
+    }
+  }, [isVirtualDrag, hasInput]);
+  let labelProps = useLabels({'aria-label': labelText, 'aria-labelledby': textId});
+
+  let {buttonProps, isPressed} = useButton({
+    onPress: () => {
+      if (inputRef.current && !hasButton  && !hasLink) {
+        inputRef.current.click();
+      }
+    }},
+    buttonRef);
+
+  let {clipboardProps} = useClipboard({
+    onPaste: (items) => props.onDrop?.({  
+      type: 'drop',
+      items,
+      x: 0,
+      y: 0,
+      dropOperation: 'copy'
+    })
+  });
+
+  let renderProps = useRenderProps({ 
+    ...props,
+    values: {isHovered, isFocused, isFocusVisible, isDropTarget, isPressed},
+    defaultClassName: 'react-aria-DropZone'
+  });
+  let DOMProps = filterDOMProps(props);
+  delete DOMProps.id;
+
+  let {pressProps} = usePress({
+    onPress: (e) => {
+      if (onPress) {
+        onPress(e);
+      }
+      if (inputRef.current && !hasButton  && !hasLink) {
+        inputRef.current.click();
+      }
+    },
+    onPressChange,
+    onPressEnd,
+    onPressStart,
+    onPressUp
+  });
+
+  let onInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    if (!e.target.files) {
+      return;
+    } else if (props.onDrop) {
+      props.onDrop({
+        type: 'drop',
+        dropOperation: 'copy',
+        x: 0,
+        y: 0,
+        items: [...e.target.files].map((file) =>
+          ({
+            kind: 'file', 
+            type: file.type, 
+            name: file.name,
+            getFile: () => Promise.resolve(file),
+            getText: () => file.text()
+          })
+        )
+      }
+      );
+    }
+  };
+  
+  return (
+    <Provider
+      values={[
+        [InputContext, {ref: inputRef, type: 'file', style: {display: 'none'}, onChange: onInputChange}],
+        [LinkContext, {slot: 'file', onPress: () => inputRef.current?.click()}], 
+        [ButtonContext, {slot: 'file', onPress: () => inputRef.current?.click()}],
+        [TextContext, {id: textId, slot: 'heading'}]
```

Comment:
I'm wondering if these slots are a bit too opinionated for RAC. A file upload button seems somewhat unrelated from dnd itself. Hiding it by default, and connecting it to Buttons and Links is also kinda specific. Maybe instead we should split this into two components: DropZone, and FileTrigger or something like that (name tbd). Then you could nest those, or use them independently.

```jsx
<FileTrigger>
  <Button /> or <Link />
</FileTrigger>
```